msprime/likelihood.py

```diff
@@ -0,0 +1,195 @@
+#
+# Copyright (C) 2019 University of Oxford
+#
+# This file is part of msprime.
+#
+# msprime is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# msprime is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with msprime.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Module responsible for computing likelihoods.
+"""
+import math
+import msprime
+import numpy as np
+
+
+def unnormalised_log_mutation_likelihood(arg, theta):
+    # log_likelihood of mutations on a given ARG up to a normalising constant
+    # that depends on the pattern of observed mutations, but not on the ARG
+    # or the mutation rate
+    tables = arg.tables
+    number_of_edges = len(tables.edges)
+    edges_above = {}
+    edge = 0
+    total_material = 0
```

Comment:
It would save a bit of typing and be more efficient to do it like this:
```python
edges_above = collections.defaultdict(list)
edges_below = collections.defaultdict(list)
time = tables.nodes.time
for j, edge in enumerate(tables.edges):   
    total_material += (edge.right - edge.left) * (time[edge.parent] - time[edge.child]])
    edges_above[edge.child].append(j)
    edges_below[edge.parent].append(j)    
```

Reply:
Also, might be easier to add the actual edge objects to these lists rather than the indexes. Would save looking up the tables again afterwards.

Reply:
I agree about storing the edge objects rather than indices. As for building both dictionaries in one loop, I don't think that would be straightforward. In general, every entry of edges_above and edges_below will be a list of multiple edges, and the behaviour of the body of the loop needs to be different based on whether the current edge will be a new entry in the dictionary, or appended to an existing one. Currently, the while-loops on lines 47 and 65 scroll through edges that don't need new entries, and append them to the correct existing entry. The loops will trigger different numbers of times for edges_below and edges_above, respectively, so combining them into one would would involve extra bookkeeping at a minimum, and probably require looking through both dictionaries at each iteration to check whether or not a given edge already has an entry.

Reply:
This is handled automatically by the defaultdict though, isn't it? C code will look quite different anyway, the aim here would be to get the cleanest, simplest Python expression of the algorithm IMO.